### DISCUSSION TITLE
Multidimensional Array | Slug | 11ty Pagination

### DISCUSSION THREAD

**Discussion by romariobertalucci** [11ty/eleventy#2354] [Uncategorized]
Hello
For example, we have next json-file

```
{
  "en": {
	"level1lang": "en",
    "categories": {
      "cat1": {
        "subcat1": {
          "id": "slug-1",
          "title": "title-1"
        },
        "subcat2": {
          "id": "slug-2",
          "title": "title-2"
        },
        "subcat3": {
          "id": "slug-3",
          "title": "title-3"
        }
      },
      "cat2": {
        "subcat1": {
          "id": "slug-1",
          "title": "title-1"
        },
        "subcat2": {
          "id": "slug-2",
          "title": "title-2"
        },
        "subcat3": {
          "id": "slug-3",
          "title": "title-3"
        }
      }
    }
  },
  "fr": {
	"level1lang": "fr",
    "categories": {
      "cat1": {
        "subcat1": {
          "id": "slug-1",
          "title": "title-1"
        },
        "subcat2": {
          "id": "slug-2",
          "title": "title-2"
        },
        "subcat3": {
          "id": "slug-3",
          "title": "title-3"
        }
      },
      "cat2": {
        "subcat1": {
          "id": "slug-1",
          "title": "title-1"
        },
        "subcat2": {
          "id": "slug-2",
          "title": "title-2"
        },
        "subcat3": {
          "id": "slug-3",
          "title": "title-3"
        }
      }
    }
  }
}
```

and we want make multilang slug's with categories, and subcategories.

ok, we put our json into _data/ folder and make njk-script, with the following content

---
...
pagination:
  data: test
  size: 1
  alias: dt
  addAllPagesToCollections: true
permalink: "{{ dt.level1lang | slug }}/"
...
---

... aaaand
we have error:
Output conflict: multiple input files are writing to `_site/undefined/index.html`

why?
because 11ty flattted our json data to
[ 'test1', 'test2' ]

so, maby fix this problem?

**Comment by pdehaan** [comment]
@romariobertalucci I think this is the relevant/interesting part:

> "Output conflict: multiple input files are writing to _site/undefined/index.html"

So your `{{ dt.level1lang }}` is actually `undefined` which causes multiple files writing to that same permalink/output path.

I think the issue here is that your src/_data/test.json file is an object and not an array, so when Eleventy loops over the object, you have to specify whether you want it to loop over the keys (default) or the object values. See https://www.11ty.dev/docs/pagination/#paging-an-object for more details.

> _If you’d like the pagination to iterate over the values instead of the keys (using `Object.values` instead of `Object.keys`), add **`resolve: values`** to your `pagination` front matter_

I think you have two possible solutions (based on my very limited understanding of your data/site):

1. if you JUST want the `level1lang`, you can simplify your `permalink` to `permalink: "{{ dt | slug }}/"` (since the `Object.keys` value is the same as your `level1lang`). This will create two files, "_site/en/index.html" and "_site/fr/index.html" and the value of `dt` is "en" and "fr" strings.
2. Set `resolve: values` in your `pagination` data, and now `dt` will be an object with all the nested properties from your test.json file.

```njk
---
# src/index.njk
pagination:
  data: test
  size: 1
  alias: dt
  addAllPagesToCollections: true
  resolve: values
permalink: "{{ dt.level1lang | slug }}/"
---

<pre>{{ dt | dump(2) | safe }}</pre>
```

### OUTPUT
```html
<!-- _site/fr/index.html -->
<pre>{
  "level1lang": "fr",
  "categories": {
    "cat1": {
      "subcat1": {
        "id": "slug-1",
        "title": "title-1"
      },
      "subcat2": {
        "id": "slug-2",
        "title": "title-2"
      },
      "subcat3": {
        "id": "slug-3",
        "title": "title-3"
      }
    },
    "cat2": {
      "subcat1": {
        "id": "slug-1",
        "title": "title-1"
      },
      "subcat2": {
        "id": "slug-2",
        "title": "title-2"
      },
      "subcat3": {
        "id": "slug-3",
        "title": "title-3"
      }
    }
  }
}</pre>
```

**Reply by romariobertalucci** [reply]
> I think the issue here is that your src/_data/test.json file is an object and not an array, so when Eleventy loops over the object, you have to specify whether you want it to loop over the keys (default) or the object values. See https://www.11ty.dev/docs/pagination/#paging-an-object for more details.

Hi, Thanx
ok, it's work for 
```
permalink: "{{ dt.level1lang | slug }}/"
```
but we can't make like this:
```
permalink: "{{ dt.level1lang | slug }}/{{ dt.categories | slug }}/"
```

**Reply by pdehaan** [reply]
Re: `permalink: "{{ dt.level1lang | slug }}/{{ dt.categories | slug }}/"`

That won't really work since `dt.categories` is a complex object and Eleventy wouldn't know how to convert this into a permalink/slug:

```json
{
    "cat1": {
      "subcat1": {
        "id": "slug-1",
        "title": "title-1"
      },
      "subcat2": {
        "id": "slug-2",
        "title": "title-2"
      },
      "subcat3": {
        "id": "slug-3",
        "title": "title-3"
      }
    },
    "cat2": {
      "subcat1": {
        "id": "slug-1",
        "title": "title-1"
      },
      "subcat2": {
        "id": "slug-2",
        "title": "title-2"
      },
      "subcat3": {
        "id": "slug-3",
        "title": "title-3"
      }
    }
  }
```

You might be able to refactor the data using [the `before` callback](https://www.11ty.dev/docs/pagination/#the-before-callback) to convert the nested object into an array of simpler objects that can be looped over. I can take a look later this evening.

Question: what are the URLs that you're expecting to create from the data?
/en/cat1/index.html?
/en/cat1/subcat1/index.html?

**Reply by romariobertalucci** [reply]
> Question: what are the URLs that you're expecting to create from the data? /en/cat1/index.html? /en/cat1/subcat1/index.html?

Yep



> You might be able to refactor the data using [the `before` callback](https://www.11ty.dev/docs/pagination/#the-before-callback) to convert the nested object into an array of simpler objects that can be looped over. I can take a look later this evening.

yes, but i'm just looking for a nicer way

for example:

```
permalink: "{{dt.level1name | slug}}/{% for t,t2 in dt.categories %}{{t2 | slug}}/{% endfor %}"
```
but, we have slug like this:
```en/cat1/cat2/```

**Reply by pdehaan** [reply]
Yeah, I don't think you'll find a nicer way, but using `before()` worked in my simple test:

```njk
---
# src/index.njk
---

<pre>{{ dt | dump(2) | safe }}</pre>
```

Then I moved all the front matter into a separate template data file so I could use the `before()` callback a bit easier in Node:
```js
// src/index.11tydata.js
module.exports = {
  pagination: {
    data: "test",
    size: 1,
    alias: "dt",
    addAllPagesToCollections: true,
    resolve: "values",
    before(data) {
      const res = [];
      for (const {level1lang, categories} of data) {
        for (const [category, subcats] of Object.entries(categories)) {
          for (const [subcategory, obj] of Object.entries(subcats)) {
            res.push({level1lang, category, subcategory, obj});
          }
        }
      }
      return res;
    },
  },
  permalink: "{{ dt.level1lang | slug }}/{{ dt.category | slug }}/{{ dt.subcategory | slug }}/"
};
```

Basically I'm building a flat array that looks something like:

```js
[
  {
    "level1lang": "en",
    "category": "cat1",
    "subcategory": "subcat1",
    "obj": { "id": "slug-1", "title": "title-1" }
  },
  ...
  {
    "level1lang": "fr",
    "category": "cat2",
    "subcategory": "subcat3",
    "obj": { "id": "slug-3", "title": "title-3" }
  }
]
```

And it seems to build the following files:

```sh
> eleventy

[11ty] Writing www/en/cat1/subcat1/index.html from ./src/index.njk
[11ty] Writing www/en/cat1/subcat2/index.html from ./src/index.njk
[11ty] Writing www/en/cat1/subcat3/index.html from ./src/index.njk
[11ty] Writing www/en/cat2/subcat1/index.html from ./src/index.njk
[11ty] Writing www/en/cat2/subcat2/index.html from ./src/index.njk
[11ty] Writing www/en/cat2/subcat3/index.html from ./src/index.njk
[11ty] Writing www/fr/cat1/subcat1/index.html from ./src/index.njk
[11ty] Writing www/fr/cat1/subcat2/index.html from ./src/index.njk
[11ty] Writing www/fr/cat1/subcat3/index.html from ./src/index.njk
[11ty] Writing www/fr/cat2/subcat1/index.html from ./src/index.njk
[11ty] Writing www/fr/cat2/subcat2/index.html from ./src/index.njk
[11ty] Writing www/fr/cat2/subcat3/index.html from ./src/index.njk
[11ty] Wrote 12 files in 0.05 seconds (4.2ms each, v1.0.1)
```
